plan/logical.go

```diff
@@ -263,18 +255,24 @@ func (v *fluxSpecVisitor) visitOperation(o *flux.Operation) error {
 
 	// no children => no successors => root node
 	if len(v.spec.Children(o.ID)) == 0 {
-		if isYield || HasSideEffect(procedureSpec) {
-			v.plan.Roots[logicalNode] = struct{}{}
-		} else {
-			// Generate a yield node
-			generateYieldNode := generateYieldNode(logicalNode)
-			err = v.addYieldName(generateYieldNode)
-			if err != nil {
-				return err
-			}
-			v.plan.Roots[generateYieldNode] = struct{}{}
-
-		}
+		// Formerly we marked nodes as roots if:
+		// - are a yield
+		// - have side effects
+		//
+		// For all other cases inside this "zero children" block, we previously
+		// added a new yield to the end of the chain and marked *the new yield*
+		// as a root.
+		// To summarize: all terminal nodes would be marked as a root and
+		// produce a result by the time the executor is done.
+		//
+		// Now, instead of inserting a yields here we are adding a result
+		// directly in the executor for all terminal nodes. Look for usages of
+		// `generateResult()` in `executor.go` to find where this happens.
+		//
+		// N.b. the number of entries in the `Roots` map factor into the
+		// concurrency quota for the query and as such we must continue to mark
+		// these all as roots here.
+		v.plan.Roots[logicalNode] = struct{}{}
```

Comment:
I'm not sure you need a description here of how it used to work. You've simplified it a lot, and one of the benefits of that is less cognitive load on folks reading the code. My hope is that no one ever needs to know how it used to work ever again. If for some reason it becomes important, we have the git history. This is just my subjective take.

Comments like these _can_ often be helpful as a comment on the PR though, so I don't want to discourage you from elucidating what you're doing; I just don't think this stuff needs to live on in perpetuity.

Reply:
I hear you. I had the same thought but no firm feeling on leaving it in or pulling it out. I'll cut or rephrase the comment to keep it current.